src/cmdexpand.c

```diff
@@ -735,94 +729,81 @@ win_redr_status_matches(
  * in "xp->xp_selected"
  */
     static char_u *
-get_next_or_prev_match(
-	int		mode,
-	expand_T	*xp)
+get_next_or_prev_match(int mode, expand_T *xp)
 {
-    int findex = xp->xp_selected;
-    int ht;
+    int	    findex;
+    int	    ht;
 
+    // When no files found, return NULL
     if (xp->xp_numfiles <= 0)
 	return NULL;
 
+    findex = xp->xp_selected;
     if (mode == WILD_PREV)
     {
-	if (findex == -1)
-	    findex = xp->xp_numfiles;
-	--findex;
+	// Select last file if at start
+	findex = (findex == -1) ? xp->xp_numfiles - 1 : findex - 1;
     }
     else if (mode == WILD_NEXT)
-	++findex;
-    else if (mode == WILD_PAGEUP)
     {
-	if (findex == 0)
-	    // at the first entry, don't select any entries
-	    findex = -1;
-	else if (findex == -1)
-	    // no entry is selected. select the last entry
-	    findex = xp->xp_numfiles - 1;
-	else
-	{
-	    // go up by the pum height
-	    ht = pum_get_height();
-	    if (ht > 3)
-		ht -= 2;
-	    findex -= ht;
-	    if (findex < 0)
-		// few entries left, select the first entry
-		findex = 0;
-	}
+	// Select next file
+	findex = findex + 1;
     }
-    else   // mode == WILD_PAGEDOWN
+    else
     {
-	if (findex == xp->xp_numfiles - 1)
-	    // at the last entry, don't select any entries
-	    findex = -1;
-	else if (findex == -1)
-	    // no entry is selected. select the first entry
-	    findex = 0;
-	else
+	// Get the height of popup menu (used for both PAGEUP and PAGEDOWN)
+	ht = pum_get_height();
+	if (ht > 3)
+	    ht -= 2;
+
+	if (mode == WILD_PAGEUP)
+	{
+	    if (findex <= 0)
+		// at the first entry or before it, handle special cases
+		findex = (findex == 0) ? -1 : xp->xp_numfiles - 1;
+	    else
+		// go up by the pum height
+		findex = (findex <= ht) ? 0 : findex - ht;
+	}
+	else    // mode == WILD_PAGEDOWN
 	{
-	    // go down by the pum height
-	    ht = pum_get_height();
-	    if (ht > 3)
-		ht -= 2;
-	    findex += ht;
-	    if (findex >= xp->xp_numfiles)
-		// few entries left, select the last entry
-		findex = xp->xp_numfiles - 1;
+	    if (findex < 0)
+		// no entry is selected, select the first entry
+		findex = 0;
+	    else if (findex >= xp->xp_numfiles - 1)
+		// at the last entry, don't select any entries
+		findex = -1;
+	    else
+		// go down by the pum height
+		findex = (findex + ht >= xp->xp_numfiles) ?
+		    xp->xp_numfiles - 1 : findex + ht;
 	}
     }
 
-    // When wrapping around, return the original string, set findex to -1.
-    if (findex < 0)
+    // Handle wrapping and boundary conditions combined
```

Comment:
What is "boundary conditions"? Isn't that the same as wrapping around?
```suggestion
    // Handle wrapping around
```